### ISSUE TITLE
Tree view initialization and opening / closing branches is slow

### ISSUE THREAD

**Issue by jjrv** [openfin/fin-hypergrid#488]
I'm using the `develop` branch of HyperGrid to display a tree with a few thousand rows and a depth of about 10. It initializes very slowly and opening or closing branches is slow (about half a second delay between clicking and the tree updating).

It seems that some hash table is not working because [this function](https://github.com/openfin/fin-hypergrid/blob/004ee4806594631dc6a4fe22589410853c69392b/src/lib/DataSourceOrigin.js#L177-L181) gets called 152876097 times at init and 22493829 times when opening or closing *any* branch (regardless of whether it's at the root, or deeper with only two leaves as children).

The tree is initialized with:

```TypeScript
import Hypergrid from 'fin-hypergrid';
import TreeView from 'fin-hypergrid/add-ons/tree-view';
import DrillDown from 'fin-hypergrid/add-ons/drill-down';
import RowById from 'fin-hypergrid/add-ons/row-by-id';
import analytics from 'hyper-analytics';

// ...

		const grid = new Hypergrid(this.node, {
			data: [{ID: null, parentID: null, key: '', value: ''}]
		});

		grid.installPlugins([
			DrillDown,
			RowById,
			['treeviewAPI', TreeView, {
				treeColumn: 'key',
				hideIdColumns: true
			}]
		]);

		grid.addProperties({
			showRowNumbers: false,
			showFilterRow: false
		});

		grid.behavior.setColumnProperties(grid.behavior.columnEnum.VALUE, {
			halign: 'left'
		});


		(window as any).fin = {
			Hypergrid: {analytics: analytics}
		};

		grid.plugins.treeviewAPI.setRelation(true);
```

All data is occasionally replaced like this (works very slowly):

```TypeScript
		grid.plugins.treeviewAPI.setRelation(false);

		grid.setData(data);
		grid.behavior.reindex();
		grid.behaviorShapeChanged();
		grid.repaint();

		grid.plugins.treeviewAPI.setRelation(true);
```

**Comment by jjrv** [comment]
Looking at the `findRow` API makes me sad:

```TypeScript
     * @param {object|null|undefined} [replacement] - One of:
     * * _omitted_ - Ignored.
     * * _object_ - Replacement for the data row if found.
     * * `null` - Flag to delete the data row if found. The found data row is nonetheless returned.
     * * `undefined` - Flag to return index of found row instead of row object itself.
```

Therefore:

`findRow(column, value, null)` and `findRow(column, value, undefined)` and `findRow(column, value)` all have different behaviour: delete the row, return its index or its contents. Reading code making such calls is depressing.

**Comment by Dwaynekj** [comment]
Hey @jjrv Thank you for your concerns, but I have to advise you to remain off the develop branch and test on alpha or the master branch. Alpha and develop may be mirrored now, but that won't always be the case (so its a future notice kind of thing). That being said, thanks for point these errors out, They shouldn't behave that way

Elaborating on my response in https://github.com/openfin/fin-hypergrid/issues/483
We will look into addressing your these bugs but also note that treeview will not be apart of core and future maintenance of it will be dropped very shortly. Abstracting it to a generic interface is the current goal in mind and fixing the bugs you pointed out would make having a general hierarchy view useful.

I'll keep you posted as I look into each of the problems mentioned

**Comment by jjrv** [comment]
The speed issue is also present in `master`, where the function that gets called millions of times is [here](https://github.com/openfin/fin-hypergrid/blob/9be421c5d086a6f091b21aeab7f2151a4db95584/src/dataSources/DataSourceOrigin.js#L146-L150).

I also have the triangle issue with the master branch (not yet sure why).

**Comment by jjrv** [comment]
I wonder what abstracting to a generic interface will mean... If I have hierarchical data like a directory tree, will showing it in a tree-like way still be supported by default, or require 3rd party add-ons?

**Comment by Dwaynekj** [comment]
Well viewing the hierarchy will be fully supported. However the restructure of the underlying data before its bound to the grid will need to be done at the application layer

**Comment by Dwaynekj** [comment]
@jjrv I'll provide an update for you this week

**Comment by jjrv** [comment]
I fixed my data so the numbers are now 10070423 calls for init and 720265 for toggling any branch. The lack of speed is no longer a blocker for my use case, but still not ideal either.

**Comment by Dwaynekj** [comment]
@jjrv 

Everyone can take a look at an review this example on the alpha branch for how to use [tree-column.html](https://github.com/openfin/fin-hypergrid/blob/develop/demo/tree-column.html) to build your own tree-view. Hypergrid's implementation of tree-view is of course going away

Please also take a look at the notes in https://github.com/openfin/fin-hypergrid/issues/569. Any feedback is appreciated.
